lib/iris/_data_manager.py

```diff
@@ -216,57 +241,76 @@ def data(self, data):
-                # Coerce input data to ndarray (including ndarray subclasses).
-                data = np.asarray(data)
-            if isinstance(data, ma.core.MaskedConstant):
-                # Promote to a masked array so that the fill-value is
-                # writeable to the data owner.
-                data = ma.array(data.data, mask=data.mask, dtype=data.dtype)
-            self._lazy_array = None
-            self._real_array = data
+            if not (hasattr(data, "shape") and hasattr(data, "dtype")):
+                data = np.asanyarray(data)
+
+            # Determine whether the class already has a defined shape,
+            # as this method is called from __init__.
+            has_shape = self._shape is not None
+            if has_shape and self.shape != data.shape:
+                # The _ONLY_ data reshape permitted is converting a 0-dimensional
+                # array i.e. self.shape == () into a 1-dimensional array of length
+                # one i.e. data.shape == (1,)
+                if self.shape or data.shape != (1,):
+                    emsg = "Require data with shape {!r}, got {!r}."
+                    raise ValueError(emsg.format(self.shape, data.shape))
+
+            # Set lazy or real data, and reset the other.
+            if is_lazy_data(data):
+                self._lazy_array = data
+                self._real_array = None
+            else:
+                if not ma.isMaskedArray(data):
+                    # Coerce input data to ndarray (including ndarray subclasses).
+                    data = np.asarray(data)
+                if isinstance(data, ma.core.MaskedConstant):
+                    # Promote to a masked array so that the fill-value is
+                    # writeable to the data owner.
+                    data = ma.array(data.data, mask=data.mask, dtype=data.dtype)
+                self._lazy_array = None
+                self._real_array = data
+            # sets ``self._shape`` if it is None, or if it is being converted from
+            # ( ) to (1, )
+            if not has_shape or (self._shape == () and data.shape == (1,)):
+                self._shape = self.core_data().shape
 
-        # Check the manager contract, as the managed data has changed.
+            # Check the manager contract, as the managed data has changed.
         self._assert_axioms()
 
     @property
     def dtype(self):
         """The dtype of the realised lazy data or the dtype of the real data."""
-        return self.core_data().dtype
+        return self.core_data().dtype if not self.is_dataless() else None
 
     @property
     def ndim(self):
         """The number of dimensions covered by the data being managed."""
-        return self.core_data().ndim
+        return len(self.shape)
 
     @property
     def shape(self):
         """The shape of the data being managed."""
-        return self.core_data().shape
+        return self._shape if self._shape else self.core_data().shape
```

Comment:
This is defensive; it wouldn't cause any issues **that I can think of** to just have `return self._shape`, but this should protect us in case I've missed something.

Reply:
I've actually removed this now. 

() is falsy. I could have easily done `...if self._shape is not None`, but I decided there's no point in adding redundant code. Easy undo if my reviewer disagrees!